utils/fork-tree/src/lib.rs

```diff
@@ -696,43 +697,56 @@ mod node_implementation {
 
 		pub fn import<F, E: std::error::Error>(
 			&mut self,
-			mut hash: H,
-			mut number: N,
-			mut data: V,
+			hash: H,
+			number: N,
+			data: V,
 			is_descendent_of: &F,
 		) -> Result<Option<(H, N, V)>, Error<E>>
 			where E: fmt::Debug,
 				  F: Fn(&H, &H) -> Result<bool, E>,
 		{
-			if self.hash == hash {
-				return Err(Error::Duplicate);
-			};
+			let mut stack: Vec<(&Self, bool)> = vec![(&self, true)];
```

Comment:
Stacking current iteration index would feel a bit more natural (and produce a smaller stack).

Reply:
I don't quite understand what do you mean by that. Why would that be useful?

Reply:
Not really useful, just that by stacking next child to visit index, you avoid stacking all children of a node.
You only ever got on stack a branch path and visit other path (taking care to apply the is_descendant condition only on node pop for depth first checking).
The indexes correspond roughly to the stored state of the iterator of at each recursive calls (one could also stack the iterator).
It is just saving a bit of stack space.

Reply:
I feel that, because these iterative functions are going to be more complicated than the recursive ones anyway, that we should be optimizing for readability instead of size.

Reply:
Readability is very subjective in this case, for instance I have a hard time following the double stack insertion of items, when stack of nodes with an index is pretty much what I am used too. But as I say it is subjective and does not matter that much.